lib/stringify.js

```diff
@@ -63,8 +67,8 @@ var stringify = function stringify( // eslint-disable-line func-name-matching
         obj = serializeDate(obj);
     }
 
-    if (obj === null) {
-        if (strictNullHandling) {
+    if (obj === null || obj === emptyValue) {
+        if (strictNullHandling || (obj === emptyValue && emptyValue !== null)) {
             return encoder && !encodeValuesOnly ? encoder(prefix, defaults.encoder, charset) : prefix;
```

Comment:
This code is sort of confusing, but I think this results in the most expected outcomes:

1. No options: `{ a: null }` becomes `'a='`
2. `strictNullHandling`: `{ a: null }` becomes `'a'`
3. `emptyValue: 'something'`: `{ a: 'something' }` becomes `'a'`
4. `emptyValue: null`: `{ a: null }` becomes `'a'`

The relevant cases for this patch are 3 and 4. Without the additional check on line 71, case 3 would give `'a='` and case 4 would give `'a='`. However to me it seems if the user provides `emptyValue` they implicitly want "strict empty handling" to be true, as I don't think returning `'a='` for 3 and 4 is ever a useful outcome.

This also means that `emptyValue: null` is equivalent to `strictNullHandling: true`.

Reply:
Hmm, I'm not sure I agree - `a=` explicitly means "empty" to me, whereas `a` could be shorthand for `a: true`. I think I might expect case 3 to become `a=`, and case 6 (`emptyValue: 'something' + `strictNullHandling` makes `{ a: 'something' }` become `a`?

Reply:
I do agree that case 3 giving `a=` is more expected, but I think it's less useful. I'm imagining the following:

1. User does `qs.parse('foo&bar=1', { emptyValue: true })` and gets `{ foo: true, bar: 1 }`
2. User increments `bar` by 1 and re-serializes: `qs.stringify({ foo: true, bar: 2 }, { emptyValue: true })`. I'd expect to get `foo&bar=2` which follows the format of the original input, not `foo=&bar=2` here.

Regardless... I do agree that requiring `strictNullHandling: true` for this behavior for `stringify` _does_ feel more logical. Returning `a` just strikes me as more in line with the use case this PR is meant to address.

Reply:
I guess the confusion is because we have an implicit empty value of null, and a separate “strict null handling” var. if it was “strict empty value handling”, then it’d be easy to make it make sense. I’m not sure what the best approach here is.

Reply:
I think it makes sense to forbid the strictNullHandling option entirely when emptyValue is not null.

However, then it may make sense to add a new option, strictEmptyHandling, that subsumes strictNullHandling in all cases?

Reply:
+1 on the first part, that makes it more consistent with `parse` behavior.

I'm honestly still unconvinced on the second part, as I still don't think anyone would ever want to use `stringify` + `emptyValue` without also wanting "strict" behavior. It just seems like it would lead to user error since the desired outcome requires `emptyValue` to always be used in tandem with `strictFooHandling`...

Reply:
Both with and without the strict option, what’s the roundtrip behavior of a null?

Reply:
```javascript
stringify({ a: null })
// => 'a='
parse('a=')
// => { a: '' }

// strict - matches
stringify({ a: null }, { strictNullHandling: true })
// => 'a'
parse('a', { strictNullHandling: true })
// => { a: null }
```

I'd expect `emptyValue: 'foobar'` to basically be `strictFoobarHandling: true`:

```javascript
stringify({ a: 'foobar' }, { emptyValue: 'foobar' })
// => 'a'
parse('a', { emptyValue: 'foobar' })
// => { a: 'foobar' }
```